gtsam/nonlinear/GncOptimizer.h

```diff
@@ -0,0 +1,381 @@
+  }
+  /// Access a copy of the parameters
+  GncParameters getParams() const {
+    return GncParameters(params_);
+  }
+  /// Access a copy of the GNC weights
+  Vector getWeights() const {
+    return weights_;
+  }
+  /// Compute optimal solution using graduated non-convexity
+  Values optimize() {
+    // start by assuming all measurements are inliers
+    weights_ = Vector::Ones(nfg_.size());
+    GaussNewtonOptimizer baseOptimizer(nfg_, state_);
+    Values result = baseOptimizer.optimize();
+    double mu = initializeMu();
+    double mu_prev = mu;
+
+    // handle the degenerate case that corresponds to small
+    // maximum residual errors at initialization
+    // For GM: if residual error is small, mu -> 0
+    // For TLS: if residual error is small, mu -> -1
+    if (mu <= 0) {
+      if (params_.verbosityGNC >= GncParameters::VerbosityGNC::SUMMARY) {
+        std::cout << "GNC Optimizer stopped because maximum residual at "
+                     "initialization is small." << std::endl;
+        result.print("result\n");
+      }
+      return result;
+    }
+
+    for (size_t iter = 0; iter < params_.maxIterations; iter++) {
+
+      // display info
+      if (params_.verbosityGNC >= GncParameters::VerbosityGNC::VALUES) {
+        result.print("result\n");
+        std::cout << "mu: " << mu << std::endl;
+        std::cout << "weights: " << weights_ << std::endl;
+      }
+      // weights update
+      weights_ = calculateWeights(result, mu);
+
+      // variable/values update
+      NonlinearFactorGraph graph_iter = this->makeWeightedGraph(weights_);
+      GaussNewtonOptimizer baseOptimizer_iter(graph_iter, state_);
+      result = baseOptimizer_iter.optimize();
+
+      // update mu
+      mu_prev = mu;
+      mu = updateMu(mu);
+
+      // stopping condition
+      if (checkMuConvergence(mu, mu_prev)) {
+        // display info
+        if (params_.verbosityGNC >= GncParameters::VerbosityGNC::SUMMARY) {
+          std::cout << "final iterations: " << iter << std::endl;
+          std::cout << "final mu: " << mu << std::endl;
+          std::cout << "final weights: " << weights_ << std::endl;
+        }
+        break;
+      }
+    }
+    return result;
+  }
+
+  /// initialize the gnc parameter mu such that loss is approximately convex (remark 5 in GNC paper)
+  double initializeMu() const {
+    // compute largest error across all factors
+    double rmax_sq = 0.0;
+    for (size_t i = 0; i < nfg_.size(); i++) {
+      if (nfg_[i]) {
+        rmax_sq = std::max(rmax_sq, nfg_[i]->error(state_));
+      }
+    }
+    // set initial mu
+    switch (params_.lossType) {
+    case GncParameters::GM:
+      return 2 * rmax_sq / params_.barcSq; // initial mu
+    case GncParameters::TLS:
+      // initialize mu to the value specified in Remark 5 in GNC paper
+      // degenerate case: residual is close to zero so mu approximately equals
+      // to -1
+      return params_.barcSq / (2 * rmax_sq - params_.barcSq);
```

Comment:
@jingnanshi : I don't think mu can be negative. does it make sense to have a lower bound for mu in this and the GM case? it the max residual is below barcsq, it means there is no outlier and maybe we can just "return"

Reply:
@lucacarlone I completely agree. I put the check outside the initializeMu function (see line 210 to 217) so that we can easily just return the current result.

Reply:
great, @jingnanshi !